partner_edit_restrict/models/res_partner.py

```diff
@@ -0,0 +1,23 @@
+# Copyright 2023 Quartile Limited
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    restrict_edit = fields.Boolean(
+        help="Enable to restrict the editing record for non-system users."
+    )
+
+    def write(self, values):
+        if self.env.user.has_group("base.group_system"):
+            return super(ResPartner, self).write(values)
+        if self.env.user._is_superuser():
+            return super(ResPartner, self).write(values)
+        for record in self:
+            if record.restrict_edit:
+                raise UserError(_("You are not allowed to modify this partner record."))
+        return super(ResPartner, self).write(values)
```

Comment:
Adding consideration on cases where multiple records are being updated.
```suggestion
        if self.env.user.has_group("base.group_system") or self.env.user._is_superuser():
            return super().write(values)
        restricted_recs = self.filtered(lambda x: x.restrict_edit)
        if restricted_recs:
            names = ", ".join(restricted_recs.mapped("name"))
            raise UserError(_("You are not allowed to modify following partner(s): %s", names)
        return super().write(values)
```